### DISCUSSION TITLE
Use only id_token_hint to check if logout confirmation screen should be displayed

### DISCUSSION THREAD

**Discussion by mposolda** [keycloak/keycloak#12183] [Ideas]
With the changes in Keycloak 18 around the support of RP-Initiated Logout, we introduced the "Logout confirmation screen", which can be displayed to the user. The logout confirmation screen is currently displayed in case that any of these is true:
- Parameter `id_token_hint` is not sent to the logout endpoint
- Client referenced by ID Token from `id_token_hint` parameter has "Consent Required" set to true

Note that for clients with "Consent Required", the logout confirmation is currently always displayed and it is not possible to disable it. This may not be ideal behaviour as some clients may want to display consent screen during login (EG. Screen "Do you really want to grant application XY access to your email and profile?"), but they may not want to display logout confirmation screen during logout. And vice-versa. For example see this issue reported: https://github.com/keycloak/keycloak/issues/11915

### Solution

The possibility is to introduce dedicated flag like "Confirm Logout" . But this will add another flag to the client to already large amount of other client options.

So better solution might be to leverage parameter `client_id` , which was added few weeks ago to the latest draft of OIDC RP-Initiated Logout https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout . This allows some flexibility as now it is possible to reference client used for the `post_logout_redirect_uri` by the `client_id` parameter. In the previous draft, there was hard requirement to send `id_token_hint` in case you send also `post_logout_redirect_uri` to be able to identify client, but `id_token_hint` is used also as a CSRF token to allow skip the logout confirmation screen. So it was not possible to have automatic redirect after logout without sending CSRF token (`id_token_hint`). Now it is possible.

So what we can do is:

- In case that valid `id_token_hint` is sent, we will always skip "Logout confirmation screen" (even if client has "Consent Required" set to true)
- In case that `id_token_hint` is not sent, we will always display logout confirmation screen. People can still have automatic redirect after logout by sending just `client_id` with `post_logout_redirect_uri`.

So if client wants user to confirm logout, it can avoid sending `id_token_hint`.

We can add option to keycloak.js adapter to `logout` method like `skipIdTokenHint`, which will cause adapter to not send `id_token_hint` to Keycloak during logout and hence enforce that logout confirmation screen is displayed. Similar for node.js adapter.

**Comment by pedroigor** [comment]
I think your proposal is also aligned with the specs, right? IIRC, there you should ask for confirmation whenever the `id_token_hint` is not given.

I agree that introducing a new option is not needed.

**Reply by mposolda** [reply]
Yes, exactly. This is aligned with the specification. Thanks Pedro!

**Reply by sschu** [reply]
I am a little late to the party, but the way I understand the spec it totally makes sense to ask the user for confirmation if only the client_id is provided while not asking when it is the id_token_hint as this means the user must have given the ID token to the client before. Just don't forget to "When both client_id and id_token_hint are present, the OP MUST verify that the Client Identifier matches the one used when issuing the ID Token." ;)

**Reply by mposolda** [reply]
@sschu Yes, good point. This is already covered in latest keycloak main branch.

**Comment by rutuparna-java** [comment]
Hello,
         to make logout we used below url
Step 1 > http://<IP>:<port>/auth/realms/apci/protocol/openid-connect/logout
Step 2 > it will ask for confirmation , Do you want to logout? 
                  Note : here if want to cancel , no such option provided.
Step 3 > if Logout then it will show msg as "You are logged out" 

But to re login again , user has to put fresh url in browser. to get login page.

also in Client options "Front Channel Logout" or/and Backchannel Logout url" mentioned and in keycloak logs can see

DEBUG [org.keycloak.services.managers.ResourceAdminManager] (default task-58) logout success for http://<ip:port>/ams/index.jsp: true

But Login did not come until page is refreshed manually. 

Thanks in Advance.

**Reply by mposolda** [reply]
Thanks for the feedback. The new option `client_id` , which is discussed here, might be exactly a help for the issues you mentioned.

If you include `client_id` like for example `http://:/auth/realms/apci/protocol/openid-connect/logout?client_id=your-client-id` and your client has "Base URL" set, then the logout confirmation page will contain link "Back to the application" and user can go back instead of confirm login. Also from the "You are logged out" page, there will be link "Back to the application". This is convenient for the cases when you don't have id_token_hint to be included.

Just a note that this will be available in Keycloak 19. In Keycloak 18, you might include ID Token from id_token_hint and/or post_logout_redirect_uri, but this is more complicated if you don't have an option to include proper ID Token when calling this endpoint.

**Reply by rutuparna-java** [reply]
Can you guide , how to fetch "id_token_hint" in java 

is this the way to get it
"Parameters
id_token_hint

When the user is redirected to the endpoint, they will be prompted if they really want to sign-out. This prompt can be bypassed by a client sending the original id_token received from authentication. This is passed as a query string parameter called id_token_hint."

**Reply by rutuparna-java** [reply]
For Keycloak version 18 
A >
By using below code in application , able to get id_token , But calling  
http://IP:Port/auth/realms/apci/protocol/openid-connect/logout?client_id=<client-id>

will ask for "do u want to logout"
will show "You are logged out"
But expected to return back to login page again (Base URL is set , still no luck).

Java Code 
request.getUserPrincipal().getKeycloakSecurityContext().getIdTokenString();

--------
with another option tried 
B > http://ip:port/auth/realms/apci/protocol/openid-connect/logout?id_token_hint=<...>&post_logout_redirect_uri=http://10.2.0.34:90/jms

This will show login page again. (Which is good )

But if we Login again , and reuse the old id_token_hint (which i printed and copied on browser 
)to logout then still it logout that user.

Is this the correct approach ?

**Comment by mposolda** [comment]
@rutuparna-java You can try latest Keycloak nightly build https://github.com/keycloak/keycloak/releases/tag/nightly . It already contains the added support for `client_id` parameter and also some fix related to `id_token_hint` . It should work in a way that:
- Parameter `client_id` should work in a way I described above (In case that your client has "Base URL" set, you should be able to see the link "Back to the application" on the logout confirmation page)
- Parameter `id_token_hint` still supported, but it must be ID Token of the current browser session. Hence not possible to reuse it.

Let me know if you see any issues with latest build.

**Reply by rutuparna-java** [reply]
We are able to do logout in one application(it will ask for confirmation ). 
But another application (which are logged in under same user)does not get logout Page nor Login Page (back channel is specified for both application).

for BackChannel, can you guide what steps need to be performed.

In Below keycloak logs we can see, user explicitly logout app2.
so app2 session id found to keycloak but not app1 sessionid.
 

org.keycloak.services.managers.AuthenticationManager] (default task-20) backchannel logout to: App1
org.keycloak.services.managers.ResourceAdminManager] (default task-20) logout resource App1 url: http://ip:port/App1/Logout.do sessionIds: 
org.keycloak.services.managers.ResourceAdminManager] (default task-20) logout success for http://ip:port/App1/Logout.do: false
org.keycloak.services.managers.AuthenticationManager] (default task-20) backchannel logout to: App2
org.keycloak.services.managers.ResourceAdminManager] (default task-20) logout resource App2 url: http://ip:port/App2 sessionIds: [zxAaXMtbuM-xvFeHIsw2EPouXa42l0lVGMckl0_j]
org.keycloak.services.managers.ResourceAdminManager] (default task-20) logout success for http://ip:port/App2: true

**Reply by rutuparna-java** [reply]
Hi @mposolda , can you guide on above points

**Comment by mposolda** [comment]
@rutuparna-java I may missed some details of your use-case and what exactly is going on. It seems that you are using Keycloak adapters in both of your applications. So it is about of configuration of the adapter, Keycloak client  and the web application of App1. You should make sure that the code-to-token request sent from App1 during login, uses the parameter `client_session_state` with the ID of the HttpSession of your "appplication session" . It seems that it does not use that in your case (not 100% sure why). If App1 does not work and App2 works, it is probably best for you to investigate the difference between them and make sure that App1 uses same/similar configuration and environment as App2.

Anyway, your isue is not directly to the "RP Initiated logout", but rather to the "Backchannel logout", so a bit unrelated to this issue.

**Comment by rutuparna-java** [comment]
Both App1 and App2 are same code/ear/war only context name is different.

**Reply by rutuparna-java** [reply]
@mposolda  
Keycloak adaptor are used in Wildfly 18 
Below steps are performed.

1 > app1 url :  http://HaProxyIp:Port/App1
    Due to Adaptor setting in Wildfly it went to Keycloak url for login.

http://HaProxyIp:Port/auth/realms/apci/protocol/openid-connect/auth?
response_type=code&client_id=app1&
redirect_uri=http://HaProxyIp:Port/App1/MainTree.do&
state=d3cff6c2-3c5a-4b13-9724-7e9f0e03a9cb&
login=true&
scope=openid

2 > After successful validation at keycloak end it opens application main page.

As u suggested added client_session_state
3 > Now in App1 , we have URL (/app2?client_session_state=<>) of another application (keycloak adaptor installed in app2 wildlfy and clint mentioned in /auth)
on this New webpage opens for App2.

4 > In App 1 or App2 , if we do Logout then the application 
which explicitly click logout gets the Login page again , but the another application does not get any call from Keycloak 
via backchannel to get again Login or Logout page.

In Logs after adding Client_session_state can see both apps got the session id

logout resource App1 url: http://IP:PORT/App1 **sessionIds:** [35GqzYw66xjqSe78qfYCxfeVBPSBhhwHj4ZE2ym0]
2022-06-13 16:40:24,226 DEBUG [org.keycloak.services.managers.ResourceAdminManager] (default task-30) logout success for http://IP:PORT/App1: **true**
2022-06-13 16:40:24,226 TRACE [org.keycloak.models.sessions.infinispan.InfinispanKeycloakTransaction] (default task-30) Adding cache operation: REPLACE on 000789f1-13dc-4bc5-b9ab-aaf28f2a2b9f
2022-06-13 16:40:24,226 TRACE [org.keycloak.models.sessions.infinispan.InfinispanKeycloakTransaction] (default task-30) Adding cache operation: REPLACE on 000789f1-13dc-4bc5-b9ab-aaf28f2a2b9f
2022-06-13 16:40:24,226 DEBUG [org.keycloak.services.managers.AuthenticationManager] (default task-30) backchannel logout to: App2
2022-06-13 16:40:24,226 TRACE [org.keycloak.keys.DefaultKeyManager] (default task-30) Active key found: realm=npci kid=odPfQzm8R0CbsLiK6yeTsKB4G31ZLkx-6BQBPjwditU algorithm=RS256 use=SIG
2022-06-13 16:40:24,227 DEBUG [org.keycloak.services.managers.ResourceAdminManager] (default task-30) logout resource App2 url: http://IP:PORT/App2 **sessionIds**: [yvGoVV9SC7BgSuW6ANTn19v9fwfslzQWTvmuMIyY]
2022-06-13 16:40:24,234 DEBUG [org.keycloak.services.managers.ResourceAdminManager] (default task-30) logout success for http://IP:PORT/App2: true


Can you guide on Point 4.

**Reply by rutuparna-java** [reply]
For Workaround , we have added javascript "refresh" code in Login.ftl , so when application makes logout and if there are many iframes like Menu, header, body page .instead of 3 login page, now due to refresh java script it will show single page.

**Comment by eliza-budeanu** [comment]
Hi, regarding the last paragraph
 
> We can add option to keycloak.js adapter to logout method like skipIdTokenHint, which will cause adapter to not send id_token_hint to Keycloak during logout and hence enforce that logout confirmation screen is displayed. Similar for node.js adapter.
 
Is there any plan for implementing it or is there a different way to show the logout confirmation page if we are using the Keycloak.js adapter?
 
Looking trough the adapter it looks like it's always attaching the id_token_hint but not sure if I'm looking at the right place :)
 
https://github.com/keycloak/keycloak/blob/1145568324d07102da5fbc1d92f5de6a122b6bef/js/libs/keycloak-js/src/keycloak.js#L494

**Comment by Flo0806** [comment]
Hello!
@mposolda, I use the oauth2-proxy with keycloak. so the complete authentication flow is handled by oauth2-proxy, configured in a NgINX server. My problem is: We use this case in our company because we handle multiple authentication with multiple providers in mutltiple apps. And it works as well. Now we  wanna update Keycloak to version 23+... and I cannot disable the logout confirm message, because all the token thinge are completly handled by oauth2-proxy... I have no access to the refresh_token, or the token_id.
If I call the ..../token endpoint I get it, but then I create a new "session" - and the logout witout confirm dialog will not working...

Do you have an idea for this problem? I mean... We all know that the user has soooo much settings. But this one more settings to disable this confirm dialog will be needed in our case..

Thx in advance and have a nice day!

Florian

**Reply by rafind** [reply]
@Flo0806 I happen to be in the same situation and have decided to go for client_id with confirmation screen. Did you managed to figure out how to get the  id_token_hint in a setup where oauth2proxy handles all the authentication flow? 

Did you find any documentation about this? Googlesearch only yields blog posts and GH issues for me :D.
